### PR TITLE
added cors support, added origin header for schedule JSON response

### PR DESCRIPTION
**Pull request by shreyas-satish**

### CODE REVIEW

File: funnel/views/helpers.py [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # -*- coding: utf-8 -*-
 
+from flask import request, abort
+from functools import wraps
+from urlparse import urlparse
+from funnel import app
```
Comment (jace) [inline funnel/views/helpers.py]: `from .. import app`

File: funnel/views/helpers.py [hunk 2]
```diff
@@ -20,3 +24,64 @@ def localize_date(date, from_tz=utc, to_tz=utc):
             to_tz = pytz_timezone(to_tz)
         return from_tz.localize(date).astimezone(to_tz).replace(tzinfo=None)
     return date
+
+def basepath(url):
```
Comment (jace) [inline funnel/views/helpers.py]: Looks like this should be in Coaster as a `root_url` utility.

File: funnel/views/helpers.py [hunk 3]
```diff
@@ -20,3 +24,64 @@ def localize_date(date, from_tz=utc, to_tz=utc):
...
+
+        basepath("https://hasgeek.com/1")
+        >> u"https://hasgeek.com
+
+    :param url: A valid URL unicode string. Eg: https://hasgeek.com
+    """
+    parsed_url = urlparse(url)
+    if not (parsed_url.scheme or parsed_url.netloc):
+        raise ValueError("Invalid URL")
+    return u"{scheme}://{netloc}".format(scheme=parsed_url.scheme, netloc=parsed_url.netloc)
```
Comment (jace) [inline funnel/views/helpers.py]: Shouldn't there be a trailing slash?

File: funnel/views/helpers.py [hunk 4]
```diff
@@ -20,3 +24,64 @@ def localize_date(date, from_tz=utc, to_tz=utc):
...
+
+def cors(any_get=False):
+    """
+    Adds CORS headers to the decorated view function.
+
+    Requires `app.config['ALLOWED_ORIGINS']` to be defined with a list
+    of permitted domains. Eg: app.config['ALLOWED_ORIGINS'] = ['https://example.com']
```
Comment (jace) [inline funnel/views/helpers.py]: We're not hardcoding this into the config. Find a way to make this dynamic.

File: funnel/views/helpers.py [hunk 5]
```diff
@@ -20,3 +24,64 @@ def localize_date(date, from_tz=utc, to_tz=utc):
...
+        >> u"https://hasgeek.com
+
+    :param url: A valid URL unicode string. Eg: https://hasgeek.com
+    """
+    parsed_url = urlparse(url)
+    if not (parsed_url.scheme or parsed_url.netloc):
+        raise ValueError("Invalid URL")
+    return u"{scheme}://{netloc}".format(scheme=parsed_url.scheme, netloc=parsed_url.netloc)
+
+def cors(any_get=False):
```
Comment (jace) [inline funnel/views/helpers.py]: This decorator should be in `coaster.views` as it's a dupe of what Boxoffice has.

File: funnel/views/helpers.py [hunk 6]
```diff
@@ -20,3 +24,64 @@ def localize_date(date, from_tz=utc, to_tz=utc):
...
+    of permitted domains. Eg: app.config['ALLOWED_ORIGINS'] = ['https://example.com']
+
+    :param any_get: Specify * on the Access-Control-Allow-Origin header for any GET request
+    """
+    def inner(f):
+        def add_headers(resp, origin):
+            resp.headers['Access-Control-Allow-Origin'] = origin
+            resp.headers['Access-Control-Allow-Methods'] = 'POST, OPTIONS, GET'
+            # echo the request's headers
+            resp.headers['Access-Control-Allow-Headers'] = request.headers.get('Access-Control-Request-Headers')
```
Comment (jace) [inline funnel/views/helpers.py]: Uh, passing back input data verbatim?

File: funnel/views/helpers.py [hunk 7]
```diff
@@ -20,3 +24,64 @@ def localize_date(date, from_tz=utc, to_tz=utc):
...
+
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            origin = request.headers.get('Origin')
+            if not origin:
+                # Firefox doesn't send the Origin header, so read the Referer header instead
+                # TODO: Remove this conditional when Firefox starts adding an Origin header
+                referer = request.referrer
+                if referer:
+                    origin = basepath(referer)
```
Comment (jace) [inline funnel/views/helpers.py]: Is Firefox's behavioral documented, and is this the recommended workaround? It's a potential security hole otherwise.

**Comment by shreyas-satish** [comment]
Use the `cors` handler from coaster. Added in https://github.com/hasgeek/coaster/pull/135.

**Pull request closed without merging**